cdragontoolbox/wad.py

```diff
@@ -93,18 +93,34 @@ def read_data(self, f):
         elif self.type == 3:
             return zstd_decompress(data)
         elif self.type == 4:
-            # Data is split into individual subchunks that are zstd compressed
-            return zstd_decompress(data)
+            # Data is split into individual subchunks that may be zstd compressed
+            if subchunkTOC is not None:
+                subchunkEntries = [struct.unpack('<IIQ', subchunkTOC[16*(subchunkIndex := self.first_subchunk_index+i):16*(subchunkIndex+1)]) for i in range(self.subchunk_count)]
+                output = bytes()
+                offset = 0
+                for subchunk in subchunkEntries:
```

Comment:
You could use `for (compressed, uncompressed, _hash) in subchunk_entries:` to at least "document" the content of the tuple (while still not bothering to create a dedicated class).